src/zir_sema.zig

```diff
@@ -1319,8 +1383,39 @@ fn validateSwitch(mod: *Module, scope: *Scope, target: *Inst, inst: *zir.Inst.Sw
     // validate for duplicate items/missing else prong
     switch (target.ty.zigTypeTag()) {
         .Enum => return mod.fail(scope, inst.base.src, "TODO validateSwitch .Enum", .{}),
-        .ErrorSet => return mod.fail(scope, inst.base.src, "TODO validateSwitch .ErrorSet", .{}),
         .Union => return mod.fail(scope, inst.base.src, "TODO validateSwitch .Union", .{}),
+        .ErrorSet => {
+            const gotten_err_set = target.ty.getErrs();
+            const is_anyerror = gotten_err_set == .anyerror;
+            var is_single = gotten_err_set == .err_single;
+            if (is_anyerror and !(inst.kw_args.special_prong == .@"else")) {
+                return mod.fail(scope, inst.base.src, "else prong required when switching on type 'anyerror'", .{});
+            }
+            var seen_values = std.HashMap(Value, usize, Value.hash, Value.eql, std.hash_map.DefaultMaxLoadPercentage).init(mod.gpa);
+            defer seen_values.deinit();
+            for (inst.positionals.items) |item| {
+                const resolved = try resolveInst(mod, scope, item);
+                const casted = try mod.coerce(scope, target.ty, resolved);
+                const val = try mod.resolveConstValue(scope, casted);
+                const err_name = val.cast(Value.Payload.Error).?.name;
+                if (try seen_values.fetchPut(val, item.src)) |prev| {
+                    return mod.fail(scope, item.src, "duplicate switch value", .{});
+                }
+                if (is_anyerror) {
+                    continue;
+                } else if (is_single and !mem.eql(u8, gotten_err_set.err_single, err_name)) {
+                    return mod.fail(scope, item.src, "expected type '{}', found '{}'", .{ gotten_err_set.err_single, gotten_err_set.err_single });
+                }
+                // TODO print this error, but it will never happen because coerce will handle it above
+                // else if (gotten_err_set == .multiple) { // we know it is an actual error set
+                //     if (gotten_err_set.multiple.get(err_name) == null) {
+                //         return mod.fail(scope, item.src, "'{}' not a member of destination error set", .{err_name});
+                //     }
+                // }
+            }
+            if (!is_single and !is_anyerror and gotten_err_set.multiple.size > inst.positionals.items.len and !(inst.kw_args.special_prong == .@"else"))
+                return mod.fail(scope, inst.base.src, "switch must handle all possibilities", .{});
```

Comment:
This should list all errors not handled as in stage1 `"error.%s not handled in switch"`.

Reply:
Ok so when trying to fix this, I am running into #7218 because this decl needs to have multiple compile errors for all the different possibilities not handled. This makes me want to get #7191 merged so that we have this functionality. The only thing blocking for me is some memory leaks that are hard for me to debug. (They are probably trivial and I am just not used to manual memory management). Would you be able to try to figure out what is going on with the memory leaks so that we can merge it? I suspect that there are other use cases than just @compileLog and this for having multiple errors for a single decl, so it would be useful anyways.

Reply:
Reverted because of discussion in #7218.